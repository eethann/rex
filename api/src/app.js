import express from 'express'

import { ApolloServer } from 'apollo-server-express'
import typeDefs from 'rex-schema'
import context from './context'
import resolvers from './schema/resolvers'

import {setupAuth} from './auth'

const app = express();

setupAuth(app)

export const apolloServer = new ApolloServer({
  typeDefs, resolvers, context
})
apolloServer.applyMiddleware({ app })

export default app
