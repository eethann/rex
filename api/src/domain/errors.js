export class UserExistsError extends Error {}
export class UsernameRequiredError extends Error {}
