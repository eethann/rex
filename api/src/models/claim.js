import Sequelize from 'sequelize'

export default (sequelize, DataTypes) => {
  var Claim = sequelize.define('Claim', {
    id: {
      allowNull: false,
      primaryKey: true,
      defaultValue: DataTypes.UUIDV1,
      type: DataTypes.UUID
    },
    url: DataTypes.TEXT,
    title: DataTypes.TEXT,
    imageUrl: DataTypes.TEXT,
    imageDescription: DataTypes.TEXT
  }, {tableName: 'claims'});
  Claim.associate = function(models) {
    Claim.belongsTo(models.User, {as: 'author'})
  };
  return Claim;
};
