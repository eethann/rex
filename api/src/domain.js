import {Claim, User, sequelize } from './models'

export const setupCleanDatabase = () => sequelize.sync({force: true})

export const clearDatabase = async () => {
  await Claim.destroy({truncate: true, cascade: true})
  await User.destroy({truncate: true, cascade: true})
}

export const cleanupDatabaseConnection = () => sequelize.close()
