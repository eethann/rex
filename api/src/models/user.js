import { UserExistsError, UsernameRequiredError } from '../domain/errors'

import securePassword from 'secure-password'
const pwd = securePassword({
  memlimit: securePassword.MEMLIMIT_DEFAULT,
  opslimit: securePassword.OPSLIMIT_DEFAULT
})


export default (sequelize, DataTypes) => {
  var User = sequelize.define('User', {
    id: {
      allowNull: false,
      primaryKey: true,
      defaultValue: DataTypes.UUIDV1,
      type: DataTypes.UUID
    },
    username: DataTypes.TEXT,
    email: DataTypes.TEXT,
    password: DataTypes.BLOB
  }, {tableName: 'users'});
  User.associate = function(models) {
    User.hasMany(models.Claim, {foreignKey: 'authorID'})
  };
  User.generateHash = (password) =>
    pwd.hash(Buffer.from(password))

  User.findByUsername = (username) =>
    User.find({where: {username}})

  User.register = (user, password) => new Promise(async (resolve, reject) => {
    const { username } = user
    if (username) {
      const existingUser = await User.findByUsername(username)
      if (!existingUser) {
        return resolve(User.create({...user, password: await User.generateHash(password)}))
      } else {
        return reject(new UserExistsError(""))
      }
    } else {
      return reject(new UsernameRequiredError(""))
    }
  })

  User.prototype.verifyPassword = function(password) {
    const self = this
    return new Promise((resolve, reject) => {
      const bufferedPassword = Buffer.from(password)
      pwd.verify(bufferedPassword, self.password, (err, code) => {
        if (err) {
          reject(err)
        } else {
          switch (code) {
          case securePassword.INVALID_UNRECOGNIZED_HASH:
            resolve(false)
          case securePassword.INVALID:
            resolve(false)
          case securePassword.VALID:
            resolve(true)
          case securePassword.VALID_NEEDS_REHASH:
            console.log('Yay you made it, wait for us to improve your safety')
            pwd.hash(bufferedPassword, function (err, improvedHash) {
              if (err) {
                console.error('You are authenticated, but we could not improve your safety this time around', err)
                resolve(true)
              } else {
                self.update({password: improvedHash}).
                  then(() => resolve(true),
                       (err) => reject(err))
              }
            })
            break
          }
        }
      })
    })
  }

  return User;
};
