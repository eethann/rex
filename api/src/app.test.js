import request from 'supertest'

import { User, sequelize } from './models'
import { setupCleanDatabase, clearDatabase, cleanupDatabaseConnection } from './domain'

import app from './app'


beforeAll(setupCleanDatabase)

afterEach(clearDatabase)

afterAll(cleanupDatabaseConnection)

expect.extend({
  toBeLoggedIn(response) {
    if (!response) {
      return {pass: false, message: "response is falsy"}
    } else if (!response.headers) {
      return {pass: false, message: "response.headers is falsy"}
    } else if (!response.headers['set-cookie']) {
      return {pass: false, message: "set-cookie header is not set"}
    } else if (!response.headers['set-cookie'][0]) {
      return {pass: false, message: "there is no cookie value in set-cookie"}
    } else {
      expect(response.headers['set-cookie'][0]).toEqual(
        expect.stringMatching(/jwt=.*; Path=.*; HttpOnly/)
      )
      return {pass: true}
    }
  }
})

test("register", async () => {
  await expect(User.findAll()).resolves.toEqual([])
  const response = await request(app).
        post("/register").type('form').
        send({username: 'travis', password: "brontosaurus"})

  expect(response.status).toBe(302)
  expect(response).toBeLoggedIn()

  const users = await User.findAll()
  expect(users[0].username).toEqual("travis")
})

describe("login", () => {
  const username = "travis"
  const password = "brontosaurus"
  beforeEach(() => User.register({username}, password))

  test("valid password", async () => {
    const response = await request(app).
          post("/login").
          type('form').send({username, password})

    expect(response.status).toBe(302)
    expect(response).toBeLoggedIn()
  })

  test("invalid password", async () => {
    const response = await request(app).
          post("/login").
          type('form').send({username, password: "HAMSNSTUFF"})

    expect(response.status).toBe(302)
    expect(response).not.toBeLoggedIn()
  })

  describe("with current-user endpoint", async () => {

    app.get("/current-user", (request, response) => {
      if (request.user) {
        response.send(JSON.stringify(request.user))
      } else {
        response.sendStatus(401)
      }
    })

    test("it 401s when not logged in", async () => {
      const response = await request(app).
            get("/current-user")
      expect(response.status).toBe(401)
    })

    test("it returns the user when logged in", async () => {
      const loginResponse = await request(app).
            post("/login").
            type('form').send({username, password})
      const response = await request(app).
            get("/current-user").
            set('Cookie', loginResponse.headers['set-cookie'])


      expect(response.status).toBe(200)
      const user = JSON.parse(response.text)
      expect(user).toMatchSnapshot({id: expect.any(String),
                                    iat: expect.any(Number),
                                    exp: expect.any(Number)
                                   })
    })
  })
})
