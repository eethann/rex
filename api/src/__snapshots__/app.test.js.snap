// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login with current-user endpoint it returns the user when logged in 1`] = `
Object {
  "exp": Any<Number>,
  "iat": Any<Number>,
  "id": Any<String>,
}
`;
